### ISSUE TITLE
Memory Management - hugetlbfs filesystem partially implemented

### ISSUE THREAD

**Issue by kphillisjr** [microsoft/WSL#3796]
The hugetlbfs filesystem implementation is incomplete. In general programs will use the path ```/sys/kernel/mm/hugepages``` to detect supported huge page sizes, and query the memory statistics of these allocations. More information on this can be found at the following links...

Additional Documentation over the topic:
* Debian Wiki page: https://wiki.debian.org/Hugepages#Huge_pages_sizes
* LWN.Net  Article over Huge Pages: https://lwn.net/Articles/376606/
* libhugetlbfs library ( https://github.com/libhugetlbfs/libhugetlbfs ) - This library is the golden standard for testing Huge page support on linux and it's subsystems.
* Current Documentation over HugePages: https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/mm/hugetlbpage.rst
*Current documentation in regards to Hugetlbfs Reservation: https://github.com/torvalds/linux/blob/master/Documentation/vm/hugetlbfs_reserv.rst
* Older Documentation over Huge pages: https://github.com/torvalds/linux/blob/e6590740ceb83fd014fae7d571fe5a5d5886b7c8/Documentation/vm/hugetlbpage.txt 

**Please fill out the below information:**
* Your Windows build number:  Microsoft Windows 10 Pro [Version 10.0.17763.253]

* What you're doing and what's happening:  The memory interfaces for /sys/kernel/mm/hugepages are missing. To test this, run the commmand...
``` grep "" -r /sys/kernel/mm/hugepages ```
The result of this command is...  
```grep: /sys/kernel/mm/hugepages: No such file or directory```

* What's wrong / what should be happening instead: This should show the available hugepage sizes. For example, If the system only supports 2MB Hugepages, The output of this command may look like this...  
```  /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages:0
/sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages_mempolicy:0
/sys/kernel/mm/hugepages/hugepages-2048kB/nr_overcommit_hugepages:0
/sys/kernel/mm/hugepages/hugepages-2048kB/free_hugepages:0
/sys/kernel/mm/hugepages/hugepages-2048kB/resv_hugepages:0
/sys/kernel/mm/hugepages/hugepages-2048kB/surplus_hugepages:0
```

* Strace of the failing command, if applicable: Not Applicable. 

* WSL launch issue: Not Applicable.

**Comment by aganea** [comment]
@kphillisjr Is that the reason why `hugetlbfs` can't be mounted? (running on 1803)
```
$ cat /proc/filesystems | grep huge
nodev   hugetlbfs
$ sudo mount -t hugetlbfs nodev /mnt/hugepages
mount: /mnt/hugepages: unknown filesystem type 'hugetlbfs'.
$ uname -a
Linux MTL-BH279 4.4.0-17134-Microsoft #523-Microsoft Mon Dec 31 17:49:00 PST 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by kphillisjr** [comment]
> 
> @kphillisjr Is that the reason why `hugetlbfs` can't be mounted? (running on 1803)
> 

Yes, but the real reason for that is that the Windows Subsystem for Linux advertises support for this filesystem type without actually implementing the filesystem. Either way, The end result of this bug is that developers ( and server administrators ) suffer performance losses due the nature of handling an increased number of page translations. 

**Step 1:** Query to see the how many Huge pages are currently available. 
```
sysctl vm.nr_hugepages
```
Ubuntu Linux Result: ( Default settings)
```
vm.nr_hugepages = 0
```
Result for Windows Subsystem for Linux: ( This shows the feature is not supported ).
```
sysctl: cannot stat /proc/sys/vm/nr_hugepages: No such file or directory
```

**Step 2:** Query to see the Default Huge page support using /proc/meminfo

```  grep Huge /proc/meminfo
```

Ubuntu Linux Result (Default Settings):
```
AnonHugePages:      108544 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
```

Windows Subsystem for Linux: ( This would be a good result if the feature was fully functional. )
```
AnonHugePages:      2048 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
```


**Step 3:** Actually allocate some Huge pages for use. For my example I will create 128 pages of the default size. Using the above example, This will set aside 256Mb of memory for use for hugepages. Also, the effects of this command will only last until the system is rebooted.
```
sudo sysctl -w vm.nr_hugepages=128
```
Ubuntu Result: 
```
vm.nr_hugepages = 128
```
Windows Subsystem for Linux response: ( this is an error due to missing features )...
```
sysctl: cannot stat /proc/sys/vm/nr_hugepages: No such file or directory
```

**Step 4:** verify the effects of this command took place in /proc/meminfo.
```
 grep Huge /proc/meminfo
```
Ubuntu Linux Result:
```
AnonHugePages:      108544 kB
HugePages_Total:       128
HugePages_Free:        128
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
```

Windows Subsystem for Linux Result: The output of this command is incorrect since the chain of commands should have result in 128 HugePages being allocated.
```
AnonHugePages:      2048 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
```
**Step 5:** The results of checking the contents of /sys/kernel/mm/hugepages/hugepages-2048kB.
```
grep "" /sys/kernel/mm/hugepages/hugepages-2048kB/*
```
Ubuntu Linux Result:
```
/sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages_mempolicy:128
/sys/kernel/mm/hugepages/hugepages-2048kB/nr_overcommit_hugepages:128
/sys/kernel/mm/hugepages/hugepages-2048kB/free_hugepages:128
/sys/kernel/mm/hugepages/hugepages-2048kB/resv_hugepages:0
/sys/kernel/mm/hugepages/hugepages-2048kB/surplus_hugepages:0
```
Windows Subsystem for Linux Result: This should mirror Ubuntu's result seeing as how the default huge page size on both platforms is 2048 kB. However, the result is an error.
```
grep: /sys/kernel/mm/hugepages/hugepages-2048kB/*: No such file or directory
```

**Comment by pkaminsk** [comment]
Any news in this field ? Would be nice to have it working at some point.

**Comment by Mr-ARB** [comment]
any news if wsl resolved ?

**Comment by mayur-who** [comment]
Any updates on this? I wanted to run/develop SPDK application and I am not able to do so without huge pages. I am stuck with slower virtualbox VM

**Comment by pervognsen** [comment]
This should probably be closed--I had no problem allocating hugepages on WSL 2. I assume WSL 1 is in maintenance mode.

**Comment by mayur-who** [comment]
any update?

**Comment by zijiwork** [comment]
any update?
